src/spac/transformations.py

```diff
@@ -102,6 +103,103 @@ def phenograph_clustering(
+        k=50,
+        output_annotation="knn",
+        associated_table=None,
+        missing_label = "no_label",
+        **kwargs):
+    """
+    Calculate knn clusters using sklearn KNeighborsClassifier
+
+    The function will add these two attributes to `adata`:
+    `.obs[output_annotation]`
+        The assigned int64 class labels by KNeighborsClassifier
+
+    `.uns["knn_features"]`
+        The features used to calculate the knn clusters
+
+    Parameters
+    ----------
+    adata : anndata.AnnData
+       The AnnData object.
+
+    features : list of str
+        The variables that would be included in creating the phenograph
+        clusters.
+    
+    annotation : str
+        The name of the annotation used for classifying the data
+
+    layer : str, optional
+        The layer to be used in calculating the phengraph clusters.
+
+    k : int, optional
+        The number of nearest neighbor to be used in creating the graph.
+
+    output_annotation : str, optional
+        The name of the output layer where the clusters are stored.
+
+    associated_table : str, optional
+        If set, use the corresponding key `adata.obsm` to calcuate the
+        clustering. Takes priority over the layer argument.
+
+    missing_label : str or int
+        The value of missing annotations in adata.obs[annotation]
+
+    Returns
+    -------
+    None
+        adata is updated inplace 
+    """
+
+    # read in data, validate annotation in the call here
+    _validate_transformation_inputs(
+        adata=adata,
+        layer=layer,
+        associated_table=associated_table,
+        features=features,
+        annotation=annotation,
+    )
+
+    if not isinstance(k, int) or k <= 0:
+        raise ValueError("`k` must be a positive integer")
+    
+    data = _select_input_features(
+        adata=adata,
+        layer=layer,
+        associated_table=associated_table,
+        features=features
+    )
+
+    # boolean mask for labeled data
+    annotation_data = adata.obs[annotation]
+    annotation_mask = annotation_data != missing_label
+    # handle np.nan values
+    annotation_mask &= pd.notnull(annotation_data)
+
+    # check that annotation is non-trivial
+    if all(annotation_mask):
+        raise ValueError("All cells are labeled. Please provide a mix of labeled and unlabeled data.")
+    elif not any(annotation_mask):
+        raise ValueError("No cells are labeled. Please provide a mix of labeled and unlabeled data.")
+
+    # fit knn classifier to labeled data and predict on full dataset     
+    data_labeled = data[annotation_mask]
+    annotation_labeled = np.array(annotation_data[annotation_mask], dtype=int)
+    
+    classifier = KNeighborsClassifier(n_neighbors = k, **kwargs)
+    classifier.fit(data_labeled, annotation_labeled)
+    knn_predict = classifier.predict(data)
+    # store output 
+    adata.obs[output_annotation] =  pd.Categorical(knn_predict)
+    adata.uns["knn_features"] = features

```

Comment:
@oeldaghar, suggestion, instead of adata.uns["knn_features"], use adata.uns[output_annotation]